src/WebJobs.Extensions.ApiHub/Listener/ApiHubListener.cs

```diff
@@ -105,24 +114,56 @@ public void Dispose()
 
             var functionResult = await _executor.TryExecuteAsync(input, CancellationToken.None);
 
+            string pollStatus = null;
+            var uri = obj as Uri;
+
+            if (uri != null)
+            {
+                pollStatus = uri.AbsoluteUri;
+            }
+            else
+            {
+                pollStatus = obj.ToString();
+            }
+
             if (functionResult.Succeeded)
             {
-                string pollStatus = null;
-                var uri = obj as Uri;
+                // If function successfully completes then the next poll Uri will be logged in an Azure Blob.
+                var status = new ApiHubStatus
+                {
+                    PollUrl = pollStatus
+                };
+
+                await SetNextPollStatusAsync(status);
+            }
+            else 
+            {
+                var status = await GetLastPollStatusAsync();
 
-                if (uri != null)
+                if (status == null)
                 {
-                    pollStatus = uri.AbsoluteUri;
+                    status = new ApiHubStatus();
                 }
-                else
+
+                if (status.RetryCount < MaxFunctionExecutionRetryCount)
                 {
-                    pollStatus = obj.ToString();
+                    status.RetryCount++;
+                    _trace.Error(string.Format("Function {0} failed to successfully proces file {1}. Number of retries: {2}. ", _functionName, apiHubFile.Path, status.RetryCount));
+
+                    await SetNextPollStatusAsync(status);
+                    await OnFileWatcher(file, obj);
                 }
+                else
+                {
+                    // The maximum retries for the function execution has reached. The file info will be added to the poison queue and the next poll status will be updated to skip the file.
+                    status.PollUrl = pollStatus;
+                    status.RetryCount = 0;
+                    _trace.Error(string.Format("Function {0} failed to successfully proces file {1} after max allowed retries of {2}. The file info will be moved to queue {3}.", _functionName, apiHubFile.Path, MaxFunctionExecutionRetryCount, PoisonQueueName));
 
-                // If function successfully completes then the next poll Uri will be logged in an Azure Blob.
-                await SetNextPollStatusAsync(pollStatus);
+                    await MoveToPoisonQueueAsync(apiHubFile);
+                    await SetNextPollStatusAsync(status);
```

Comment:
If the status update fails after you've moved to the poison queue, you could get duplicates in the queue right? I realize these two operations can't be done as an atomic transaction.